en/ros2/px4_ros2_interface_lib.md

```diff
@@ -0,0 +1,365 @@
+};
+```
+
+- `[1]`: First we create a class that inherits from `px4_ros2::ModeBase`
+- `[2]`: In the constructor, we pass the mode name. This also allows us to configure some other things, like replacing a flight controller internal mode.
+- `[3]`: This is where we create all objects that we want to use later on.
+  This can be RC input, setpoint type(s), or telemetry. `*this` is passed as a `Context` to each object, which associates the object with the mode.
+- `[4]`: Whenever the mode is active, this method gets called regularly (the update rate depends on the setpoint type).
+  Here is where we can do our work and generate a new setpoint.
+
+After creating an instance of that mode, `mode->doRegister()` must be called which does the actual registration with the flight controller and returns `false` if it fails.
+In case a mode executor is used, `doRegister()` must be called on the mode executor, instead of for the mode.
+
+### Setpoint Types
+
+A mode can choose its setpoint type(s) it wants to use to control the vehicle.
+The used types also define the compatibility with different vehicle types.
+
+The following sections provide a list of commonly used setpoint types.
+You can also add your own type by adding a class that inherits from `px4_ros2::SetpointBase`, sets the configuration flags according to what the setpoint requires, and then publishes any topic containing a setpoint.
+
+<!--
+#### GoTo Position Setpoints
+
+TODO
+-->
+
+#### Direct Actuator Control Setpoint (DirectActuatorsSetpointType)
+
+Actuators can be directly controlled using the [px4_ros2::DirectActuatorsSetpointType](https://github.com/Auterion/px4-ros2-interface-lib/blob/main/px4_ros2_cpp/include/px4_ros2/control/setpoint_types/direct_actuators.hpp) setpoint type.
+Motors and servos can be set independently.
+Be aware that the assignment is vehicle and setup-specific.
+For example to control a quadrotor, you need to set the first 4 motors according to its [output configuration](../concept/control_allocation.md).
+
+:::note
+If you want to control an actuator that does not control the vehicle's motion, but for example a payload servo, see [below](#controlling-an-independent-actuator-servo).
+:::
+
+### Controlling an Independent Actuator/Servo
+
+If you want to control an independent actuator (a servo), follow these steps:
+
+1. [Configure the output](../payloads/#generic-actuator-control-with-mavlink)
+2. Create an instance of [px4_ros2::PeripheralActuatorControls](https://github.com/Auterion/px4-ros2-interface-lib/blob/main/px4_ros2_cpp/include/px4_ros2/control/peripheral_actuators.hpp) in the constructor of your mode.
+3. Call the `set` method to control the actuator(s).
+   This can be done independently of any active setpoints.
+
+### Telemetry
+
+Telemetry, such as local or global position estimates can be found under [px4_ros2/odometry](https://github.com/Auterion/px4-ros2-interface-lib/tree/main/px4_ros2_cpp/include/px4_ros2/odometry).
```

Comment:
Are these topics the same as the normal PX4 ones? Would be good to explain the origin etc.

Reply:
Yes this will need a lot more documentation.

Reply:
I'll add links to the ones we know exist now.